src/Functions/parseDateTime.cpp

```diff
@@ -0,0 +1,1747 @@
+            {
+                date.reset();
+
+                StringRef str_ref = col_str->getDataAt(i);
+                Pos cur = str_ref.data;
+                Pos end = str_ref.data + str_ref.size;
+                for (const auto & instruction : instructions)
+                {
+                    // std::cout << "instruction:" << instruction.toString() << std::endl;
+                    cur = instruction.perform(cur, end, date);
+                    // std::cout << "date:" << date.toString() << std::endl;
+                }
+
+                // Ensure all input was consumed.
+                if (cur < end)
+                    throw Exception(
+                        ErrorCodes::LOGICAL_ERROR,
+                        "Invalid format input {} is malformed at {}",
+                        str_ref.toView(),
+                        std::string_view(cur, end - cur));
+
+                Int64 time = date.checkAndGetDateTime(*time_zone);
+                data_res.push_back(static_cast<UInt32>(time));
+            }
+
+            return col_res;
+        }
+
+
+    private:
+        class Action
+        {
+        private:
+            using Func = std::conditional_t<
+                parse_syntax == ParseDateTimeTraits::ParseSyntax::MySQL,
+                Pos (*)(Pos, Pos, Date &),
+                std::function<Pos(Pos, Pos, Date &)>>;
+            Func func{};
+            std::string func_name;
+
+            std::string literal;
+
+        public:
+            explicit Action(Func && func_, const char * func_name_) : func(std::move(func_)), func_name(func_name_) { }
+
+            explicit Action(const String & literal_) : literal(literal_) { }
+            explicit Action(String && literal_) : literal(std::move(literal_)) { }
+
+            /// For debug
+            [[maybe_unused]] String toString() const
+            {
+                if (func)
+                    return "func:" + func_name;
+                else
+                    return "literal:" + literal;
+            }
+
+            Pos perform(Pos cur, Pos end, Date & date) const
+            {
+                if (func)
+                    return func(cur, end, date);
+                else
+                {
+                    ensureSpace(cur, end, literal.size(), "required literal size not matched");
+                    if (std::string_view(cur, literal.size()) != literal)
+                        throw Exception(
+                            ErrorCodes::LOGICAL_ERROR, "Expect literal {} but {} provided", literal, std::string_view(cur, literal.size()));
+                    cur += literal.size();
+                    return cur;
+                }
+            }
+
+            template <typename T, bool check_space = true>
+            static Pos readNumber2(Pos cur, Pos end, T & res)
+            {
+                if constexpr (check_space)
+                    ensureSpace(cur, end, 2, "readNumber2 requires size >= 2");
```

Comment:
This message will be hard to understand for users. Better include the current format string, e.g. `%c`. If necessary, include it from outside.

Reply:
I think we can easily get the whole context from stack trace of current exception.  Introducing an extra argument may slow down performance for `readNumber2` is a frequently called function.